bbbc021_feature_reduction_tests.py

```diff
@@ -0,0 +1,247 @@
+        io.imsave(name, image)
+
+
+def output_features(ims, filenames, out_file):
+    """
+    Build a default feature map for each image in ims and output a dataframe of
+    [filenames, features] to out_file as csv for reading in
+
+    :param ims: opened nparray images
+    :param filenames: filenames corresponding to each image in ims with relative path to top level directory
+    :param out_file: name of CSV file to save dataframe, with relative path to top level directory
+    """
+    # generate filenames column to exist as first column of feature DF
+    filenames_col = ["Filenames"]
+    filenames_col.extend(filenames)
+    filenames_col = pd.DataFrame(filenames_col)
+
+    all_image_features = pd.DataFrame()
+    # set up flag to only add header row once
+    flag = True
+    i = 1
+    for im, im_name in zip(ims, filenames):
+        image_features, feature_names = default_feature_map(im, sample_size=100)
+        # make sure header row is added to dataframe in first iteration
+        if flag:
+            all_image_features = all_image_features.append(pd.DataFrame(feature_names).transpose())
+            flag = False
+        image_features = pd.DataFrame(image_features).transpose()
+        all_image_features = all_image_features.append(image_features, ignore_index=True)
+        print("Image processed: {} out of {}".format(i, len(filenames)))
+        i += 1
+    # concatenate filenames column to the features and save to CSV.
+    all_image_features = pd.concat([filenames_col, all_image_features], axis=1)
+    all_image_features.to_csv(out_file)
+
+
+def pca_transform(features_filename):
+    """
+    Read a file of image features into dataframe and perform a 2 component PCA, returning the 2 component values
+    of each image
+
+    :param features_filename: filename of CSV containing image features
+    :return coords: np array of 2 components for each image
+    """
+    all_image_features = pd.read_csv(features_filename)
+    pca = PCA(2)
+    coords = pca.fit_transform(all_image_features.iloc[1:, 2:])
+
+    return coords
+
+def tsne_transform(features_filename):
+    """
+    Read a file of image features into dataframe and perform TSNE transformation, returning the 2 component values
+    of each image
+
+    :param features_filename: filename of CSV containing image features
+    :return coords: np array of 2 components for each image
+    """
+    all_image_features = pd.read_csv(features_filename)
+
+    return TSNE().fit_transform(all_image_features.iloc[1:, 2:])
+
+
+def umap_transform(features_filename):
+    """
+    Read a file of image features into dataframe and perform umap transformation, returning the 2 component values
+    of each image
+
+    :param features_filename: filename of CSV containing image features
+    :return coords: np array of 2 components for each image
+    """
+    all_image_features = pd.read_csv(features_filename)
+
+    return umap.UMAP().fit_transform(all_image_features.iloc[1:, 2:])
```

Comment:
Why are you discarding the first row?

Reply:
Because currently the dataframe column headings are being written as though they were a row of data, instead of column names. That's ok, it's been noted elsewhere.